model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/OllamaDefaultToolsHandler.java

```diff
@@ -0,0 +1,132 @@
+        List<ToolSpecification> extendedList = new ArrayList<>(toolSpecifications.size() + 1);
+        extendedList.addAll(toolSpecifications);
+        extendedList.add(DEFAULT_RESPONSE_TOOL);
+        Prompt prompt = DEFAULT_SYSTEM_TEMPLATE.apply(
+                Map.of("tools", Json.toJson(extendedList)));
+
+        // TODO handle -> toolThatMustBeExecuted skipped for the moment
+        String initialSystemMessages = messages.stream().filter(cm -> cm.role() == Role.SYSTEM)
+                .map(Message::content)
+                .collect(Collectors.joining("\n"));
+
+        Message groupedSystemMessage = Message.builder()
+                .role(Role.SYSTEM)
+                .content(initialSystemMessages + "\n" + prompt.text())
+                .build();
+
+        List<Message> otherMessages = messages.stream().filter(cm -> cm.role() != Role.SYSTEM).toList();
+
+        // Add specific tools message
+        List<Message> messagesWithTools = new ArrayList<>(messages.size() + 1);
+        messagesWithTools.add(groupedSystemMessage);
+        messagesWithTools.addAll(otherMessages);
+
+        builder.messages(messagesWithTools);
+
+        return builder;
+    }
+
+    @Override
+    public AiMessage getAiMessageFromResponse(ChatResponse response, List<ToolSpecification> toolSpecifications) {
+        ToolResponse toolResponse;
+        try {
+            // Extract tools
+            toolResponse = Json.fromJson(response.message().content(), ToolResponse.class);
+        } catch (Exception e) {
+            // No tools found
+            return AiMessage.from(response.message().content());
+        }
+        // If the tool is the final result with default response tool
+        if (toolResponse.tool.equals(DEFAULT_RESPONSE_TOOL.name())) {
+            return AiMessage.from(toolResponse.tool_input.get("response").toString());
+        }
+        // Check if tool is part of the available tools
+        List<String> availableTools = toolSpecifications.stream().map(ToolSpecification::name).toList();
+        if (!availableTools.contains(toolResponse.tool)) {
+            return AiMessage.from(String.format(
+                    "The LLM wants to call a tool %s that is not part of the available tools %s",
+                    toolResponse.tool, availableTools));
+        }
+        // Extract tools request from response
+        List<ToolExecutionRequest> toolExecutionRequests = toToolExecutionRequests(toolResponse, toolSpecifications);
+        return AiMessage.aiMessage(toolExecutionRequests);
+    }
+
+    record ToolResponse(String tool, Map<String, Object> tool_input) {
+    }
+
+    private List<ToolExecutionRequest> toToolExecutionRequests(ToolResponse toolResponse,
+            List<ToolSpecification> toolSpecifications) {
+        return toolSpecifications.stream()
+                .filter(ts -> ts.name().equals(toolResponse.tool))
+                .map(ts -> toToolExecutionRequest(toolResponse, ts))
+                .toList();
```

Comment:
We generally try hard to avoid lambdas in Quarkus code

Reply:
why (just curious)?

Reply:
When Quarkus started, the team found that the lambdas had a small (but not zero) impact on memory usage.

Mind you, this on Java 8, so things may have changed substantially since then, but we still try to avoid them unless the alternative is just plain terrible.